File: agent/sdk/typescript/src/native-program-call.ts
export interface NativeProgramCall {
  readonly programId: Uint8Array;
  readonly guestAbi: 1 | 2;
  readonly entrypoint: string;
  readonly calldata: Uint8Array;
  readonly capabilities: Uint8Array;
  readonly accessDeclaration: Uint8Array;
  readonly responseCapacity: number;
  readonly resources: readonly [bigint, bigint, bigint, bigint, bigint, bigint, bigint];
}

export function encodeNativeProgramCall(call: NativeProgramCall): Uint8Array {
  if (call.programId.length !== 32 || call.programId.every(value => value === 0)
    || (call.guestAbi !== 1 && call.guestAbi !== 2) || !/^[A-Za-z0-9_.]{1,128}$/.test(call.entrypoint)
    || call.calldata.length > 1_048_576 || call.capabilities.length > 65_535
    || call.accessDeclaration.length > 1_048_576 || !Number.isInteger(call.responseCapacity)
    || call.responseCapacity < 0 || call.responseCapacity > 1_048_576
    || call.resources.length !== 7 || call.resources.some(value => value < 0n || value > 0xffff_ffff_ffff_ffffn)) {
    throw new TypeError("invalid native program call");
  }
  const entrypoint = new TextEncoder().encode(call.entrypoint);
  const size = 106 + entrypoint.length + call.calldata.length + call.capabilities.length + call.accessDeclaration.length;
  const output = new Uint8Array(size);
  const view = new DataView(output.buffer);
  output.set(call.programId); view.setUint16(32, call.guestAbi); view.setUint16(34, entrypoint.length);
  view.setUint32(36, call.calldata.length); view.setUint16(40, call.capabilities.length);
  view.setUint32(42, call.accessDeclaration.length); view.setUint32(46, call.responseCapacity);
  call.resources.forEach((value, index) => view.setBigUint64(50 + index * 8, value));
  let offset = 106;
  for (const body of [entrypoint, call.calldata, call.capabilities, call.accessDeclaration]) {
    output.set(body, offset); offset += body.length;
  }
  return output;
}

export function decodeNativeProgramCall(payload: Uint8Array): NativeProgramCall {
  if (payload.length < 106) throw new TypeError("invalid native program call");
  const view = new DataView(payload.buffer, payload.byteOffset, payload.byteLength);
  const lengths = [view.getUint16(34), view.getUint32(36), view.getUint16(40), view.getUint32(42)];
  if (106 + lengths.reduce((total, length) => total + length, 0) !== payload.length) throw new TypeError("invalid native program call");
  let offset = 106;
  const body = (length: number): Uint8Array => { const result = payload.slice(offset, offset + length); offset += length; return result; };
  const call: NativeProgramCall = {
    programId: payload.slice(0, 32), guestAbi: view.getUint16(32) as 1 | 2,
    entrypoint: new TextDecoder("utf-8", { fatal: true }).decode(body(lengths[0]!)),
    calldata: body(lengths[1]!), capabilities: body(lengths[2]!), accessDeclaration: body(lengths[3]!),
    responseCapacity: view.getUint32(46),
    resources: [view.getBigUint64(50), view.getBigUint64(58), view.getBigUint64(66), view.getBigUint64(74), view.getBigUint64(82), view.getBigUint64(90), view.getBigUint64(98)],
  };
  encodeNativeProgramCall(call);
  return call;
}

File: agent/sdk/python/layerx_sdk/native_program_call.py
from __future__ import annotations

from dataclasses import dataclass
import re
import struct


@dataclass(frozen=True)
class NativeProgramCall:
    program_id: bytes
    guest_abi: int
    entrypoint: str
    calldata: bytes
    capabilities: bytes
    access_declaration: bytes
    response_capacity: int
    resources: tuple[int, int, int, int, int, int, int]


def encode_native_program_call(call: NativeProgramCall) -> bytes:
    if (len(call.program_id) != 32 or call.program_id == bytes(32)
            or isinstance(call.guest_abi, bool) or call.guest_abi not in (1, 2)
            or re.fullmatch(r"[A-Za-z0-9_.]{1,128}", call.entrypoint) is None
            or len(call.calldata) > 1_048_576 or len(call.capabilities) > 65_535
            or len(call.access_declaration) > 1_048_576
            or isinstance(call.response_capacity, bool) or not isinstance(call.response_capacity, int)
            or not 0 <= call.response_capacity <= 1_048_576 or len(call.resources) != 7
            or any(isinstance(value, bool) or not isinstance(value, int) or not 0 <= value < 1 << 64 for value in call.resources)):
        raise ValueError("invalid native program call")
    entrypoint = call.entrypoint.encode("ascii")
    return (struct.pack(">32sHHI HII 7Q", call.program_id, call.guest_abi, len(entrypoint),
                        len(call.calldata), len(call.capabilities), len(call.access_declaration),
                        call.response_capacity, *call.resources)
            + entrypoint + call.calldata + call.capabilities + call.access_declaration)


def decode_native_program_call(payload: bytes) -> NativeProgramCall:
    if len(payload) < 106:
        raise ValueError("invalid native program call")
    fields = struct.unpack(">32sHHI HII 7Q", payload[:106])
    lengths = fields[2:6]
    if 106 + sum(lengths) != len(payload):
        raise ValueError("invalid native program call")
    offset = 106
    bodies = []
    for length in lengths:
        bodies.append(payload[offset:offset + length])
        offset += length
    call = NativeProgramCall(fields[0], fields[1], bodies[0].decode("ascii"),
                             bodies[1], bodies[2], bodies[3], fields[6], fields[7:])
    encode_native_program_call(call)
    return call
